From d5ad6ee3ce26443ef13164a12727a95663be87aa Mon Sep 17 00:00:00 2001
From: Aaradhana Sahu <quic_aarasahu@quicinc.com>
Date: Mon, 17 Oct 2022 16:25:35 +0530
Subject: [PATCH] ath12k: fix rmmod issue in SSR case

Currently, we are calling ath12k_hif_power_down immediate after
ath12k_core_stop function, ath12k_hif_power_down is responsible for
sending mhi_power_off state to firmware by this it's power off one
chip while the core does not stop for other parallel running chips,
due to which other chips receive MHI_CB_EE_RDDM callback.

Hence all other chips are crashing during rmmod

Fix it by calling ath12k_hif_power_down function after stopping core
for all chips.

Signed-off-by: Aaradhana Sahu <quic_aarasahu@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.c b/drivers/net/wireless/ath/ath12k/core.c
index 981443f..cee7a21 100644
--- a/drivers/net/wireless/ath/ath12k/core.c
+++ b/drivers/net/wireless/ath/ath12k/core.c
@@ -1586,11 +1586,18 @@ static void ath12k_core_hw_group_stop(struct ath12k_hw_group *ag)
 		ath12k_core_pdev_destroy(ab);
 		ath12k_core_stop(ab);
 
+		mutex_unlock(&ab->core_lock);
+	}
+
+	for (i = ag->num_chip - 1; i >= 0; i--) {
+		ab = ag->ab[i];
+
+		mutex_lock(&ab->core_lock);
+
 		ath12k_hif_power_down(ab);
 
 		mutex_unlock(&ab->core_lock);
 	}
-
 	ath12k_mac_destroy(ag);
 
 	ath12k_core_hw_group_destroy(ag);
-- 
2.17.1

